core/base/v7/inc/ROOT/RError.hxx

```diff
@@ -0,0 +1,230 @@
+   explicit RResultType(const T &value) : fValue(value) {}
+   T fValue;
+};
+
+class RResultTypeVoid {};
+} // namespace Internal
+
+
+// clang-format off
+/**
+\class ROOT::Experimental::RResult
+\ingroup Base
+\brief The class is used as a return type for operations that can fail; wraps a value of type T or an RError
+
+The RResult enforces checking whether it contains a valid value or an error state. If the RResult leaves the scope
+unchecked, it will throw an exception.  RResult should only be allocated on the stack, which is helped by deleting the
+new operator.  RResult is movable but not copyable to avoid throwing exceptions multiple times.
+*/
+// clang-format on
+template <typename T>
+class RResult : public std::conditional_t<std::is_void<T>::value, Internal::RResultTypeVoid, Internal::RResultType<T>> {
+private:
+   /// This is the nullptr for an RResult representing success
+   std::unique_ptr<RError> fError;
+   /// This is safe because checking an RResult is not a multi-threaded operation
+   mutable bool fIsChecked{false};
+
+public:
+   /// Constructor is _not_ explicit in order to allow for `return T();` for functions returning RResult<T>
+   /// Only available if T is not void
+   template <typename Dummy = T, typename = typename std::enable_if_t<!std::is_void<T>::value, Dummy>>
+   RResult(const Dummy &value) : Internal::RResultType<T>(value) {}
+   /// Constructor is _not_ explicit such that the RError returned by R__FAIL can be converted into an RResult<T>
+   /// for any T
+   RResult(const RError &error) : fError(std::make_unique<RError>(error)) {}
+
+   /// RResult<void> has a default contructor that creates an object representing success
+   template <typename Dummy = T, typename = typename std::enable_if_t<std::is_void<T>::value, Dummy>>
+   RResult() {}
+
+   RResult(const RResult &other) = delete;
+   RResult(RResult &&other) = default;
+   RResult &operator =(const RResult &other) = delete;
+   RResult &operator =(RResult &&other) = default;
+
+   ~RResult() noexcept(false)
+   {
+      if (R__unlikely(fError && !fIsChecked)) {
+         // Prevent from throwing if the object is deconstructed in the course of stack unwinding for another exception
+#if __cplusplus >= 201703L
+         if (std::uncaught_exceptions() == 0)
+#else
+         if (!std::uncaught_exception())
+#endif
+         {
+            throw RException(*fError);
+         }
+      }
+   }
+
+   /// Only available if T is not void
+   template <typename Dummy = T>
+   typename std::enable_if_t<!std::is_void<T>::value, const Dummy &>
+   Get()
+   {
+      if (R__unlikely(fError)) {
+         fError->AppendMessage(" (invalid access)");
+         throw RException(*fError);
+      }
+      return Internal::RResultType<T>::fValue;
+   }
+
+   explicit operator bool() const
+   {
+      fIsChecked = true;
+      return !fError;
+   }
+
+   RError *GetError() { return fError.get(); }
+
+   void Throw() { throw RException(*fError); }
```

Comment:
Perhaps we can discuss this, I think I don't quite understand how or why `ThrowOutlined` would behave different  from the `Throw()` method.

Reply:
Clarified verbally as "user might compile user.cxx with `-fno-exceptions` and then they cannot `#include` RError.hxx as that contains `throw`. Instead, move `throw` into ROOT source file (i.e. "outlined" in contrast to "inlined")

Reply:
> and then they cannot #include RError.hxx

Do mean 'can not' as in "compilation will fail" or 'should not' as in "at run-time the code will do the 'wrong' thing?

It seems that if the user is compile the code with ```-fno-exceptions``` it would be 'surprising' if it throws exception, no?